File: src/ts/BaseStateMachine.ts
import { Machine, Message, State, Transition } from './state-machine-types';
import { v4 as uuidv4 } from 'uuid';
import { createHash } from 'crypto';

export class BaseStateMachine<TData extends any, TState extends State> implements Machine<TData, TState> {
    name: string;
    superMachine: string;
    subMachines: Machine<TData, TState>[];
    location: string;
    currentState: Transition<State>;
    processing: boolean;
    queue: PriorityQueue<Transition<State>>;
    currentModel: TData;
    log: Resolved<TData, TState>[];

    private csrfToken: string;
    private requestTokens: Map<string, string>;

    constructor(name: string, initialData: TData) {
        this.name = name;
        this.superMachine = '';
        this.subMachines = [];
        this.location = 'local';
        this.currentModel = initialData;
        this.processing = false;
        this.queue = new PriorityQueue<Transition<State>>();
        this.log = [];
        this.csrfToken = `csrf:${uuidv4()}`;
        this.requestTokens = new Map();
    }

    protected generateRequestToken(): string {
        const token = `request:${uuidv4()}`;
        this.requestTokens.set(token, new Date().toISOString());
        return token;
    }

    protected generateHash(data: any, salt: string): string {
        return createHash('sha256')
            .update(JSON.stringify(data) + salt)
            .digest('hex');
    }

    async sendMessage(address: string, data: TData): Promise<void> {
        const [machinePath, action] = address.split('/').filter(Boolean);
        const requestToken = this.generateRequestToken();
        
        const message: Message<TData, TState> = {
            from: this.currentState.from,
            to: this.currentState.to,
            send: new Date(),
            received: undefined,
            data,
            csrfToken: this.csrfToken,
            requestId: `request:${uuidv4()}`,
            requestToken,
            id: uuidv4(),
            salt: uuidv4(),
            hash: this.generateHash(data, this.csrfToken)
        };

        // Handle GraphQL sub-queries in the address
        if (machinePath.includes('~')) {
            const [basePath, subQuery] = machinePath.split('~');
            // Process sub-query if needed
            if (subQuery) {
                // Handle GraphQL sub-query resolution
                await this.resolveSubQuery(subQuery, data);
            }
        }

        // Queue the transition
        this.queue.enqueue({
            from: this.currentState.from,
            to: this.currentState.to,
            send: new Date(),
            received: undefined
        });

        // Process the queue if not already processing
        if (!this.processing) {
            await this.processQueue();
        }
    }

    private async resolveSubQuery(subQuery: string, data: TData): Promise<void> {
        // Implement GraphQL sub-query resolution logic
        // This would integrate with your GraphQL schema and resolvers
    }

    private async processQueue(): Promise<void> {
        this.processing = true;
        while (!this.queue.isEmpty()) {
            const transition = this.queue.dequeue();
            if (transition) {
                await this.executeTransition(transition);
            }
        }
        this.processing = false;
    }

    private async executeTransition(transition: Transition<State>): Promise<void> {
        // Implement transition execution logic
        // This would handle state changes and side effects
    }
} 

File: src/kotlin/core/BaseStateMachine.kt
package com.logview.core

import java.time.Instant
import java.util.UUID
import java.security.MessageDigest
import java.util.concurrent.PriorityBlockingQueue
import kotlinx.serialization.Serializable

@Serializable
abstract class BaseStateMachine<TData : Any, TState : State> {
    abstract val name: String
    abstract val superMachine: String
    abstract val subMachines: List<BaseStateMachine<TData, TState>>
    abstract val location: String
    
    var currentState: Transition<TState>? = null
    var processing: Boolean = false
    val queue = PriorityBlockingQueue<Transition<TState>>()
    var currentModel: TData? = null
    val log = mutableListOf<Resolved<TData, TState>>()
    
    protected val csrfToken: String = "csrf:${UUID.randomUUID()}"
    protected val requestTokens = mutableMapOf<String, String>()
    
    protected fun generateRequestToken(): String {
        val token = "request:${UUID.randomUUID()}"
        requestTokens[token] = Instant.now().toString()
        return token
    }
    
    protected fun generateHash(data: TData, salt: String): String {
        val messageDigest = MessageDigest.getInstance("SHA-256")
        val dataBytes = (data.toString() + salt).toByteArray()
        val hashBytes = messageDigest.digest(dataBytes)
        return hashBytes.joinToString("") { "%02x".format(it) }
    }
    
    suspend fun sendMessage(address: String, data: TData) {
        val (machinePath, action) = address.split("/").filter { it.isNotEmpty() }
        val requestToken = generateRequestToken()
        
        val message = Message(
            id = UUID.randomUUID().toString(),
            csrfToken = csrfToken,
            requestToken = requestToken,
            requestId = "request:${UUID.randomUUID()}",
            salt = UUID.randomUUID().toString(),
            hash = generateHash(data, csrfToken),
            data = data,
            from = currentState?.from ?: State.Initial(),
            to = currentState?.to ?: State.Processing(),
            send = Instant.now()
        )
        
        // Handle GraphQL sub-queries
        if (machinePath.contains("~")) {
            val (basePath, subQuery) = machinePath.split("~", limit = 2)
            if (subQuery.isNotEmpty()) {
                resolveSubQuery(subQuery, data)
            }
        }
        
        // Queue the transition
        currentState?.let { state ->
            queue.offer(Transition(
                from = state.from,
                to = state.to,
                send = Instant.now()
            ))
        }
        
        // Process the queue if not already processing
        if (!processing) {
            processQueue()
        }
    }
    
    private suspend fun resolveSubQuery(subQuery: String, data: TData) {
        // Implement GraphQL sub-query resolution
        // This would integrate with your GraphQL schema and resolvers
    }
    
    private suspend fun processQueue() {
        processing = true
        while (queue.isNotEmpty()) {
            val transition = queue.poll()
            if (transition != null) {
                executeTransition(transition)
            }
        }
        processing = false
    }
    
    protected abstract suspend fun executeTransition(transition: Transition<TState>)
} 